### PR TITLE
Bugfix Savegame Sorting and Refreshing (fixes #343, #351)

### LINKED ISSUE #343: Sort order of saved games.
With the new GUI (*`new_swing_gui` branch*), the list of saved maps is no longer in antichronological order. Instead, no sorting seems to be being applied at all, map order is the same as in `ls -f`.

### PR DESCRIPTION
**Pull request by andreas-eberle**
fixes #343, #351

### CODE REVIEW

**Comment by andreas-eberle** [comment]
@JKatzwinkel: Can you have a look at the changes?

File: jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java [hunk 1]
```diff
@@ -103,22 +111,23 @@ public OpenPanel(final List<MapLoader> maps, IMapSelectedListener mapSelectedLis
...
+	public OpenPanel(final List<? extends MapLoader> maps, final IMapSelectedListener mapSelectedListener,
+			final ListCellRenderer<MapLoader> cellRenderer) {
 		setMapLoadersWithoutSearchChanged(maps);
 		setLayout(new BorderLayout());
 
 		initFilter();
 
 		this.searchTextField = new SearchTextField();
-		searchTextField.putClientProperty(ELFStyle.KEY, ELFStyle.TEXT_DEFAULT);
+		this.searchTextField.putClientProperty(ELFStyle.KEY, ELFStyle.TEXT_DEFAULT);
```
Comment (michaelzangl) [inline jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java]: Can we decide on a global this/super policy for the whole project?

File: jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java [hunk 2]
```diff
@@ -84,12 +79,25 @@
 	/**
 	 * Constructor
 	 *
+	 * @param doubleclickListener
+	 *            Gets called when an entry is double clicked, can be <code>null</code>
+	 */
+	public OpenPanel(final ChangingList<? extends MapLoader> maps, IMapSelectedListener mapSelectedListener) {
+		this(maps.getItems(), mapSelectedListener);
+		maps.setListener(changedLister -> {
+			this.setMapLoaders(changedLister.getItems());
```
Comment (michaelzangl) [inline jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java]: this.methodname

File: jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java [hunk 3]
```diff
@@ -103,22 +111,23 @@ public OpenPanel(final List<MapLoader> maps, IMapSelectedListener mapSelectedLis
 	 * @param cellRenderer
 	 *            Cell renderer to use
 	 */
-	public OpenPanel(final List<MapLoader> maps, final IMapSelectedListener mapSelectedListener, final ListCellRenderer<MapLoader> cellRenderer) {
+	public OpenPanel(final List<? extends MapLoader> maps, final IMapSelectedListener mapSelectedListener,
+			final ListCellRenderer<MapLoader> cellRenderer) {
 		setMapLoadersWithoutSearchChanged(maps);
 		setLayout(new BorderLayout());
 
 		initFilter();
```
Comment (michaelzangl) [inline jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java]: No this.methodname

File: jsettlers.logic/src/jsettlers/logic/map/MapLoader.java [hunk 4]
```diff
@@ -83,13 +83,13 @@ public static boolean isExtensionKnown(String filename) {
...
-	public int compareTo(MapLoader o) {
-		MapFileHeader myHeader = header;
-		MapFileHeader otherHeader = o.header;
-		if (myHeader.getType() == MapType.SAVED_SINGLE) {
-			return -myHeader.getCreationDate().compareTo(otherHeader.getCreationDate()); // order by date descending
+	public int compareTo(MapLoader other) {
+		MapFileHeader myHeader = getFileHeader();
+		MapFileHeader otherHeader = other.getFileHeader();
+		if (myHeader.getType() == otherHeader.getType() && myHeader.getType() == MapType.SAVED_SINGLE) {
+			return -this.getCreationDate().compareTo(other.getCreationDate()); // order by date descending
```
Comment (michaelzangl) [inline jsettlers.logic/src/jsettlers/logic/map/MapLoader.java]: this.methodname

File: jsettlers.logic/src/jsettlers/logic/map/MapLoader.java [hunk 5]
```diff
@@ -83,13 +83,13 @@ public static boolean isExtensionKnown(String filename) {
 
 	// - Interface: Comparable<MapLoader>
 	@Override
-	public int compareTo(MapLoader o) {
-		MapFileHeader myHeader = header;
-		MapFileHeader otherHeader = o.header;
-		if (myHeader.getType() == MapType.SAVED_SINGLE) {
-			return -myHeader.getCreationDate().compareTo(otherHeader.getCreationDate()); // order by date descending
+	public int compareTo(MapLoader other) {
+		MapFileHeader myHeader = getFileHeader();
```
Comment (michaelzangl) [inline jsettlers.logic/src/jsettlers/logic/map/MapLoader.java]: no this.methodname

File: jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java (hunk 2 above)
Comment (andreas-eberle) [inline jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java]: What do you mean?

File: jsettlers.logic/src/jsettlers/logic/map/MapLoader.java (hunk 4 above)
Comment (andreas-eberle) [inline jsettlers.logic/src/jsettlers/logic/map/MapLoader.java]: Here the this.methodName() is to distinguish from other.methodName().

File: jsettlers.logic/src/jsettlers/logic/map/MapLoader.java (hunk 5 above)
Comment (andreas-eberle) [inline jsettlers.logic/src/jsettlers/logic/map/MapLoader.java]: Here, I forgot it.

File: jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java (hunk 1 above)
Comment (andreas-eberle) [inline jsettlers.main.swing/src/jsettlers/main/swing/menu/openpanel/OpenPanel.java]: In general, I don't want to enforce writing this everywhere. I prefer writing it, where I need it and if I write it once in a method, I write it everywhere in that method... I'm open for discussion on this (maybe in an issue to create a wiki entry for this).

**Comment by andreas-eberle** [comment]
@michaelzangl: I changed the code to use no this. where not necessary or good for readability (e.g. compareTo)

**Comment by michaelzangl** [comment]
@andreas-eberle In graphics, I have the policy to write this only where it is needed. And I try to only need it for constructor/setter parameter assignments. I try to not shadow any field with a local variable.

But having it in compareTo is also fine. This is why I don't want to enforce a policy everywhere.

**Comment by andreas-eberle** [comment]
OK, I think we can agree on this idea.

**Comment by JKatzwinkel** [comment]
Looks very elegantly put to me, and map sorting methods work as well as updating the list of saved games.

**Pull request merged**